File: app/web/lib/panels.ts
import type { Panel, PanelResolveRequest, PanelSummary } from './backend'

const API_BASE_URL = process.env.NEXT_PUBLIC_API_BASE_URL?.replace(/\/$/, '') ?? ''

export type PanelRequestErrorCode = 'not_found' | 'validation' | 'unavailable' | 'request_failed'

/**
 * Safe, presentation-ready failure from the panel API boundary. Response bodies
 * are deliberately not copied into this error: a gene panel drives which
 * variants a run reports on, so a failure must surface as an explicit
 * unavailable state rather than as text that could read like panel content.
 */
export class PanelRequestError extends Error {
  readonly code: PanelRequestErrorCode
  readonly status?: number

  constructor(message: string, options: { code: PanelRequestErrorCode; status?: number }) {
    super(message)
    this.name = 'PanelRequestError'
    this.code = options.code
    this.status = options.status
    Object.setPrototypeOf(this, PanelRequestError.prototype)
  }
}

function panelErrorForStatus(status: number): PanelRequestError {
  if (status === 404) return new PanelRequestError('That gene panel was not found.', { code: 'not_found', status })
  if (status === 422) {
    return new PanelRequestError('That panel request was not valid.', { code: 'validation', status })
  }
  if (status === 503) {
    return new PanelRequestError('The panel service is unavailable.', { code: 'unavailable', status })
  }
  return new PanelRequestError('The panel request failed.', { code: 'request_failed', status })
}

/** Wraps transport failures so a dropped connection cannot be mistaken for an
 *  empty or successful catalogue. */
async function requestPanels<T>(path: string, init?: RequestInit): Promise<T> {
  let response: Response
  try {
    response = await fetch(`${API_BASE_URL}${path}`, init)
  } catch (err) {
    if (err instanceof DOMException && err.name === 'AbortError') throw err
    throw new PanelRequestError('The panel service could not be reached.', { code: 'unavailable' })
  }
  if (!response.ok) throw panelErrorForStatus(response.status)
  return (await response.json()) as T
}

export async function getPanels(): Promise<PanelSummary[]> {
  const data = await requestPanels<{ panels: PanelSummary[] }>('/api/v1/panels')
  return data.panels
}

export async function getPanel(slug: string): Promise<Panel> {
  return requestPanels<Panel>(`/api/v1/panels/${encodeURIComponent(slug)}`)
}

export async function resolvePanel(input: PanelResolveRequest): Promise<Panel> {
  return requestPanels<Panel>('/api/v1/panels/resolve', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(input),
  })
}

File: app/backend/app/api/routes/panels.py
from __future__ import annotations

from fastapi import APIRouter, HTTPException, Request, status

from app.schemas.panels import Panel, PanelListResponse, PanelResolveRequest
from app.services.panels import PanelService

router = APIRouter(prefix="/api/v1/panels", tags=["panels"])


@router.get("", response_model=PanelListResponse)
def list_panels(request: Request) -> PanelListResponse:
    return _service(request).list_panels()


@router.post("/resolve", response_model=Panel)
def resolve_panel(payload: PanelResolveRequest, request: Request) -> Panel:
    return _service(request).resolve(payload)


@router.get("/{slug}", response_model=Panel)
def get_panel(slug: str, request: Request) -> Panel:
    panel = _service(request).get_panel(slug)
    if panel is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Panel '{slug}' was not found.",
        )
    return panel


def _service(request: Request) -> PanelService:
    service = getattr(request.app.state, "panel_service", None)
    if service is None:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail="Panel service is unavailable.",
        )
    return service
